### PR TITLE
[GlobalIsel][AArch64] Legalize G_SCMP and G_UCMP

### PR DESCRIPTION
**Pull request by tschuett**
https://github.com/llvm/llvm-project/pull/91871
https://github.com/llvm/llvm-project/pull/98774

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/GlobalISel/legalize-threeway-cmp.mir [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  bb.0.entry:
+    ; CHECK-LABEL: name: test_scmp
+    ; CHECK: [[COPY:%[0-9]+]]:_(s64) = COPY $x0
+    ; CHECK-NEXT: [[COPY1:%[0-9]+]]:_(s64) = COPY $x0
+    ; CHECK-NEXT: [[ICMP:%[0-9]+]]:_(s32) = G_ICMP intpred(slt), [[COPY]](s64), [[COPY1]]
+    ; CHECK-NEXT: [[ICMP1:%[0-9]+]]:_(s32) = G_ICMP intpred(sgt), [[COPY]](s64), [[COPY1]]
+    ; CHECK-NEXT: [[C:%[0-9]+]]:_(s32) = G_CONSTANT i32 1
+    ; CHECK-NEXT: [[C1:%[0-9]+]]:_(s32) = G_CONSTANT i32 0
+    ; CHECK-NEXT: [[SELECT:%[0-9]+]]:_(s32) = G_SELECT [[ICMP1]](s32), [[C]], [[C1]]
+    ; CHECK-NEXT: [[C2:%[0-9]+]]:_(s32) = G_CONSTANT i32 3
```
Comment (tschuett) [inline llvm/test/CodeGen/AArch64/GlobalISel/legalize-threeway-cmp.mir:15]: The `3` should be a `-1`. Are we printing unsigned?

Comment (nikic) [reply llvm/test/CodeGen/AArch64/GlobalISel/legalize-threeway-cmp.mir:15]: This is because your tests incorrectly use G_ANYEXT instead of G_SIGNEXT.

Comment (tschuett) [reply llvm/test/CodeGen/AArch64/GlobalISel/legalize-threeway-cmp.mir:15]: I tried G_SEXT. No differences. The G_CONSTANTs are used by the G_SELECTs. They are printed before the G_EXT.

Comment (tschuett) [reply llvm/test/CodeGen/AArch64/GlobalISel/legalize-threeway-cmp.mir:15]: G_ANYEXT is just make the type larger to copy it into an AArch64 register.

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp [hunk 2]
```diff
@@ -7269,6 +7272,36 @@ LegalizerHelper::LegalizeResult LegalizerHelper::lowerMinMax(MachineInstr &MI) {
...
+  CmpInst::Predicate LTPredicate = Cmp->isSigned()
+                                       ? CmpInst::Predicate::ICMP_SLT
+                                       : CmpInst::Predicate::ICMP_ULT;
+  CmpInst::Predicate GTPredicate = Cmp->isSigned()
+                                       ? CmpInst::Predicate::ICMP_SGT
+                                       : CmpInst::Predicate::ICMP_UGT;
+
+  auto One = MIRBuilder.buildConstant(DstTy, 1);
+  auto Zero = MIRBuilder.buildConstant(DstTy, 0);
+  auto MinusOne = MIRBuilder.buildConstant(DstTy, -1);
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:7292]: Sink these down to just before first use?

**Review by arsenm — APPROVED** [review]

**Comment by arsenm** [comment]
Title should probably say lower or legalize, not select

File: llvm/test/CodeGen/AArch64/GlobalISel/legalize-threeway-cmp.mir (hunk 1 above)
Comment (nikic) [reply llvm/test/CodeGen/AArch64/GlobalISel/legalize-threeway-cmp.mir:15]: I'd still prefer it if you changed these to G_SEXT to make the returned result correct.

Comment (arsenm) [reply llvm/test/CodeGen/AArch64/GlobalISel/legalize-threeway-cmp.mir:15]: 3 is a 2-bit -1, so the returned result is correct. The high bits don't matter

Comment (nikic) [reply llvm/test/CodeGen/AArch64/GlobalISel/legalize-threeway-cmp.mir:15]: Yes, the high bits don't matter if you use ANYEXT, but the use of ANYEXT here does not make sense semantically...

**Pull request merged**